File: 분리 집합, 최소 신장 트리/G4.1976 여행가자/조연주/Main.java
// 15852 KB, 136 ms
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N;
    static int[] parent;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine()); // 도시 수

        parent = new int[N+1];
        for(int i=1; i<=N; i++){
            parent[i] = i;
        }

        int M = Integer.parseInt(br.readLine()); // 여행 계획

        // 도시 연결정보
        for(int i=1; i<=N; i++){
            st = new StringTokenizer(br.readLine());
            for(int k=1; k<=N; k++){
                if(Integer.parseInt(st.nextToken())==1){
                    union(i,k);
                }
            }
        }
        
        // 여행 계획
        int[] plan = new int[M];
        st = new StringTokenizer(br.readLine());
        for(int i=0; i<M; i++){
            plan[i] = Integer.parseInt(st.nextToken());
        }

        // 계산
        for(int i=0; i<M-1; i++){
            if(find(plan[i])!=find(plan[i+1])){
                System.out.println("NO");
                return;
            }
        }
        System.out.println("YES");
        
    }

    static void union(int a, int b){
        a = find(a);
        b = find(b);

        if(a>b){
            parent[a] = b;
        }
        else{
            parent[b] = a;
        }
    }

    static int find(int n){
        if(parent[n]==n){
            return n;
        }
        return parent[n] = find(parent[n]);

    }
}

File: 분리 집합, 최소 신장 트리/G4.1976 여행가자/조연주/main.cc
// 메모리: 2024 KB, 시간: 0 ms
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
typedef pair<int,pair<int, int>> pp;

int parent[201];
int plan[1001];

int find(int a) {
	if (parent[a] == a) {
		return a;
	}
	return parent[a] = find(parent[a]);
}

void uni(int a, int b) {
	a = find(a);
	b = find(b);
	parent[b] = a;
}

int main() {
	ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

	int N, M;
	cin >> N >> M;

	// 부모 노드 초기화
	for (int i = 1; i <= 200; i++) {
		parent[i] = i;
	}

	// 도시 연결
	for (int i = 1; i <= N; i++) {
		for (int k = 1; k <= N; k++) {
			int n;
			cin >> n;
			if (n == 1) {
				uni(i, k);
			}
		}
	}

	// 여행 루트
	for (int i = 0; i < M; i++) {
		int n;
		cin >> n;
		plan[i] = n;
	}

	// 계산
	for (int i = 0; i < M - 1; i++) {
		if (find(plan[i]) != find(plan[i + 1])) {
			cout << "NO";
			return 0;
		}
	}
	cout << "YES";

	return 0;
}
